src/solvers/eigen_sparse_linear_solver.C

```diff
@@ -101,47 +101,109 @@ EigenSparseLinearSolver<T>::solve (SparseMatrix<T> & matrix_in,
 
   std::pair<unsigned int, Real> retval(0,0.);
 
-  const int max_its = this->get_int_solver_setting("max_its", m_its);
-  const double abs_tol = this->get_real_solver_setting("abs_tol", tol);
+  // Eigen doesn't give us a solver base class?  We'll just use a
+  // generic lambda, then.
+  auto do_solve = [this, &rhs, &solution, tol, m_its]
+    (auto & e_solver, std::string_view msg) {
+    const int max_its = this->get_int_solver_setting("max_its", m_its);
+    const double abs_tol = this->get_real_solver_setting("abs_tol", tol);
+
+    e_solver.setMaxIterations(max_its);
+    e_solver.setTolerance(abs_tol);
+    libMesh::out << msg << std::endl;
+
+    solution._vec = e_solver.solveWithGuess(rhs._vec,solution._vec);
+
+    libMesh::out << "#iterations: " << e_solver.iterations() << " / " << max_its << std::endl;
+    libMesh::out << "estimated error: " << e_solver.error() << " / " << abs_tol << std::endl;
+    _comp_info = e_solver.info();
+    return std::make_pair(e_solver.iterations(), e_solver.error());
+  };
+
+  using Eigen::DiagonalPreconditioner;
+  using Eigen::IdentityPreconditioner;
+  using Eigen::IncompleteCholesky;
+  using Eigen::IncompleteLUT;
 
   // Solve the linear system
   switch (this->_solver_type)
     {
       // Conjugate-Gradient
     case CG:
       {
-        Eigen::ConjugateGradient<EigenSM> solver (matrix._mat);
-        solver.setMaxIterations(max_its);
-        solver.setTolerance(abs_tol);
-        solution._vec = solver.solveWithGuess(rhs._vec,solution._vec);
-        libMesh::out << "#iterations: " << solver.iterations() << std::endl;
-        libMesh::out << "estimated error: " << solver.error() << std::endl;
-        retval = std::make_pair(solver.iterations(), solver.error());
-        _comp_info = solver.info();
+        const int UPLO = Eigen::Lower|Eigen::Upper;
+
+        switch (this->_preconditioner_type)
+        {
+          case IDENTITY_PRECOND:
+            {
+              Eigen::ConjugateGradient<EigenSM,UPLO,IdentityPreconditioner> solver (matrix._mat);
+              retval = do_solve(solver, "Eigen CG solver without preconditioning");
+              break;
+            }
+          case ICC_PRECOND:
+          case ILU_PRECOND: // If asked for CG we must be symmetric, right?
```

Comment:
I would not have this much trust in a user...

Reply:
It's not so much "I trust the user", it's "If the user's already screwed up this badly then we're not going to screw them up any worse".  But sure, I'll change that case to an error.